pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java

```diff
@@ -39,7 +39,11 @@
 public final class PhysicalColumnIndexContainer implements ColumnIndexContainer {
   private static final Logger LOGGER = LoggerFactory.getLogger(PhysicalColumnIndexContainer.class);
 
-  private final Map<IndexType, IndexReader> _readersByIndex;
+  private static final IndexReader[] EMPTY_READERS = new IndexReader[0];
+
```

Comment:
This change should reduce average memory size from 300B+ to ~60B.

Reply:
I like the idea of this change, but the current implementation makes this class much more complex than required. Could we create a custom class that implements this _map_ even if it doesn't actually implement java.util.Map interface? 

Even if it is an internal class here, it would make the code easier to maintain, but I think it may be common to have maps indexed by index type and given these index types are calculated at java static time, we can offer this `IndexTypeMap` class as a poor man's EnumMap

Reply:
While some of the code I added is related to missing exception handling (that could lead to memory leaks), I'd say this class is still quite small and trivial.
Is there another place we could realistically use this `map` on ? TBH I don't think map is the right abstraction here anyway. Considering that a column won't really have that many indexes at the same time, a pair of arrays (short|byte and IndexReader) and linear search could be better. 
EnumMap is cool but would take several times more memory.

Reply:
Extracted to internal static class.